### ISSUE TITLE
Support RFC8693: Token Exchange

### ISSUE THREAD

**Issue by adaniline** [ory/hydra#1218]
In a zero trust microservices architecture, it is extremely useful to have an ability to generate "on behalf of" access tokens with shorter lifespan and narrower scope when performing service to service calls.
Even though OAuth 2.0 Token Exchange (https://tools.ietf.org/html/draft-ietf-oauth-token-exchange-16) is a draft right now, it would be extremely useful if Hydra supported a way to exchange an access token with an "on behalf of" token.

**Describe the solution you'd like**
Standard-based:
- IEFT OAuth 2.0 Token Exchange draft https://tools.ietf.org/html/draft-ietf-oauth-token-exchange-16

Proprietary:
- Azure AD "On Behalf Of" flow with jwt-bearer grant https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow

**Comment by aeneasr** [comment]
While it looks like a reasonable approach, there are so many different grant extensions to the existing OAuth2 protocol in draft (or self-defined by e.g. Google) that we simply lack the resources to implement and maintain them, which is why drafts are generally not being implemented.

In my personal view, I'm not a fan of using OAuth2, a framework that was initiated to give 3rd party developers a way to access personal information, for things like workload/service auth. I think systems like [SPIFFE](https://spiffe.io/) take a much more sane approach to this. We are actually working with [Scytale](https://www.scytale.io/), co-founded by Evan Gilman (he wrote the Book [Zero Trust Networks](https://www.amazon.de/Zero-Trust-Networks-Building-Untrusted/dp/1491962194)), to deliver a way to solve workload auth in the context of OAuth2. But this will (obviously) take a while.

Closing this, for now, as a wontfix/wontdo. Feel free to keep the conversation going though.

**Comment by aberasarte** [comment]
Sorry @aeneasr for writing in this closed issue but I'm just wondering if you could give more details regarding this statement:

> In my personal view, I'm not a fan of using OAuth2, a framework that was initiated to give 3rd party developers a way to access personal information, for things like workload/service auth

We are already using Hydra and OIDC for user authentication and we are planning to use it for authorizing third party applications using the `client_credentials` grant. I thought that was the right approach and it would simplify our backend authn/authz infrastructure but maybe I'm missing something here?

**Comment by aeneasr** [comment]
You can use client_credentials if you want to. The proposed draft above is something completely different.

**Comment by aberasarte** [comment]
OK, I misunderstood your sentence then. Thanks for the clarification!

**Comment by smeir** [comment]
Hi @aeneasr 
Since January 2020 the OAuth 2.0 token exchange grant type specification is not a draft any more. 
In our projects we use Hydra as authorisation server and for some use cases we will need a token exchange. 
Are there any plans to implement this extension grant type or is there any documentation on how to do it by ourself?

**Comment by aeneasr** [comment]
We currently lack use cases / popular demand and resources to tackle this, but do welcome contributions. As a word of caution, this will be a lot of work to implement!

**Comment by 0ka** [comment]
Hi @aeneasr 
We would like to thrive this topic forward with a contribution. Digging through the source code some points are pretty clear. But we would like to have a talk before starting this to get the right direction/architecture decisions.

**Comment by aeneasr** [comment]
Sounds good! I'll reopen that - for clarification, this will be about implementing https://tools.ietf.org/html/rfc8693 right?

**Comment by 0ka** [comment]
Yes that is RFC8693.
We would create a new handler/oauth2/flow_token_exchange.go in the fosite project. Do we need an own issue in fosite or is this issue here sufficient?

**Comment by aeneasr** [comment]
Sorry, I overlooked your comment. Yes - no need to create another issue!

**Comment by mpnunes** [comment]
**Our use case:**
We have three applications A, B, and C.
User U can grant application A access to B with auth code grant. Offline access should also work, means that A will get a refresh token.
Now application A wants to access application B and B should have access to application C (in the name of User U), with offline access. Here we need the token exchange (A gets access to B in the name of U, B exchange the token to get an own access and refresh token to access C in the name of U)


**Implementation idea:**
* New flow_token_exchnage.go in fosite
* Clients that have grant type "urn:ietf:params:oauth:grant-type:token-exchange" are allowed to ask for token exchange
* Request for token_exchange is similar to client_cresentials request with additional parameters as per RFC8693 (e.g. "subject_token", "subject_token_type")


**Permission checks**
**Option 1:**
* Clients have an additional claim "may_act" with a list of all IDs of the clients that should be allowed to exchange tokens. Tokens/request from the cient than contains the "may_act" field. During processing of the token exchange fosite verify that the original token/request allows the exchange for the current client requesting the exchange. (Issue here is that the spec define only one may_act field and not a list, whould be needed for chained delegation)
**Option 2:**
* On token-exchange hydra calls a configured callback similar to the consent app that allows to make checks externaly.

What's your opinion on that?

**Comment by aeneasr** [comment]
Thank you for your thoughts! I think that sounds reasonable! The idea of `may_act` is to restrict who can exchange an access token? So in your example, the client of C would have `may_act: B`?

Are there best practices around this? I do think that it might be tricky if every client is allowed to exchange token with one another so we probably need some type of restriction.

I think a callback wouldn't work that well because there is no user interaction (similar to client_credentials), so probably option 1 would be the way to go.

**Comment by aeneasr** [comment]
Marked as stale in error.

**Comment by thomasvargiu** [comment]
Hi, any news on this?

**Comment by drwatsno** [comment]
https://datatracker.ietf.org/doc/html/rfc8693 declares new grant type "urn:ietf:params:oauth:grant-type:token-exchange". We tested hydra 2.0 for use with new grant type but without success. Also, we didn't found any configuration options for this flow or any code in codebase which is intended to support it

FYI @aeneasr

**Comment by rverma-nsl** [comment]
Same here, we are also trying to use the same. Although same is implemented in Keycloak. We were very enthusiastically planning to migrate from keycloak but unfortunately got stuck here.

**Comment by ghenry** [comment]
Why not implement it and submit a pull request?

I'm sure the project would greatly appreciate it.

**Comment by alee792** [comment]
@ghenry, the release notes state that this RFC is implemented, but I can't seem to find any trace of it in the docs or codebase.

**Comment by ghenry** [comment]
Where are folks seeing this mentioned?

Nothing on the code base - https://github.com/ory/hydra/search?q=RFC8693

**Comment by ghenry** [comment]
Nothing here https://github.com/ory/hydra/releases/tag/v2.0.2

**Comment by alee792** [comment]
https://github.com/ory/hydra/releases/tag/v2.0.0

**Comment by ghenry** [comment]
It looks like it might just be from the go client sdk:

https://github.com/ory/hydra/blob/d768cf6580b3410f7d0b3b9420760ce0818a5fe2/client/client.go#L74
https://github.com/ory/hydra/pull/3337
https://github.com/ory/hydra/blob/74d4569a0c15d7b2d1bbf435937d0cf98175ff57/internal/httpclient/docs/OAuth2TokenExchange.md
https://github.com/ory/hydra/blob/d768cf6580b3410f7d0b3b9420760ce0818a5fe2/oauth2/handler.go#L779
https://github.com/ory/hydra/blob/74d4569a0c15d7b2d1bbf435937d0cf98175ff57/internal/httpclient/docs/OAuth2Api.md
https://github.com/ory/hydra/blob/74d4569a0c15d7b2d1bbf435937d0cf98175ff57/internal/httpclient/docs/OAuth2Api.md#Oauth2TokenExchange
https://github.com/ory/hydra/blob/c586e035bb78752747845f6a27189121d2d53034/internal/httpclient/api/openapi.yaml#L3393
https://github.com/ory/hydra/blob/c586e035bb78752747845f6a27189121d2d53034/spec/swagger.json#L2889

**Comment by alee792** [comment]
I believe those are auth code for token exchange but will take a second
look.

**Comment by alee792** [comment]
A low effort search to check if the IETF URN for exchange is used within Hydra or Fosite yields no results: https://github.com/search?l=&q=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Atoken-exchange+repo%3Aory%2Fhydra+repo%3Aory%2Ffosite&type=code

@aeneasr, any chance we can get clarification and whether it's implemented or the release notes jumped the gun?

**Comment by mutexlock** [comment]
[OAuth 2.0 Token Exchange (RFC8693)](https://datatracker.ietf.org/doc/html/rfc8693) is now fully supported, including the JSON Web Token profile!

it seems only support rfc7523 @aeneasr

**Comment by aeneasr** [comment]
Yes, we messed up in here unfortunately. Can you help us identify all the places where the incorrect line is? We have to update it :)

**Comment by vinckr** [comment]
Hello all, 

have a look at this community project: https://github.com/Exact-Realty/ts-hydra-rfc8693

This will help you use OAuth2 Token Exchange until its implemented in Ory Hydra

**Comment by etodanik** [comment]
Well, that's a bit of an unfair way to categorize it. This is one of the most commented open issues. The fact that people are respectfully not spamming it doesn't mean that it's not one of the more wanted issues right now.

**Comment by aeneasr** [comment]
> The motivation for this automation is to help prioritize issues in the backlog and not ignore, reject, or belittle anyone.
>
> If this issue was marked as stale erroneously you can exempt it by adding the backlog label, assigning someone, or setting a milestone for it.

**Comment by shadiramadan** [comment]
I also require this feature.  In my use case I trust/verify a Google token in AppScript (for a looker studio connector) and want to exchange it for an access token for auth to my apis.

I’m shopping for auth solutions because I’m currently implementing oauth2 myself and that seemed like re-inventing the wheel.  This use case is a dealbreaker though.

**Comment by ashish3805** [comment]
Eagerly Waiting, This is a deal breaker.

**Comment by etodanik** [comment]
> > I mean, on my end: this issue was the reason I chose not to go with ory (and am happily using another stack for this)
> 
> Good for you! Doesn't solve this issue any faster though ;) The automation will still go in and every couple of months mark it as stale.
> 
> So far we had no commercial interest (and thus incentive) to build this. If a company / buyer comes up that needs this as a requirement we can happily put it on the roadmap.

If you don't want to know why someone decided to pay your competitors and not you - don't ask (in the GitHub automation message). I was just providing information that seemed to have been requested (that this was a deal breaker issue in adopting this vs the competition)

**Comment by dushantCars24** [comment]
Token exchange is standard flow now in any Auth provider tool.

**Comment by gmile** [comment]
We're already using Ory for OAuth2, and it would be nice to have support for RFC8693.

The use-case I have in mind is authentication between an MCP server and a resource server (e.g. product API). Today, after MCP client sends authentication information to MCP server, the MCP protocol currently neither clarifies nor recommends what shall happens next authentication-wise, only saying the `auth_tokens` coming from clients can't be used to pass them through to the resource service:

> The MCP server MUST NOT pass through the token it received from the MCP client.
> 
> [link](https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization)

In this situation, it would be convenient to verify `auth_token` with IdP (Hydra) and exchange it for another one, this time suited for a server-to-server interaction, e.g. to talk to the resource server API. This is where having the "exchange of tokens" (e.g. RFC8693) implemented in Ory could help a lot.

In the mean-time, projects like https://github.com/ApelegHQ/ts-hydra-rfc8693 could serve as a work-around, though it appears as if they would require deploying & maintaining a separate service.
